### PR TITLE
Map

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

File: test/fixture/mapTest.scrypt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+
+contract MapTest {
+    static const int MAX = 10;
+
+    // @state
+    bytes _mpData;
+
+    public function unlock(int key, int val, int keyIndex)// var length array
```
Comment (xhliu) [inline test/fixture/mapTest.scrypt:15]: Add some branches to test. Also inside loops. E.g.,

```
loop () {
  if () {
    map.set()
  } else {
    map.delete()
  }
}
```

File: src/utils.ts [hunk 2]
```diff
@@ -1405,4 +1405,97 @@ export function deserializeArgfromASM(contract: AbstractContract, arg: Argument,
   }
 
   arg.value = value;
+}
+
+function flattenData(data: SingletonParamType): ScryptType[] {
```
Comment (xhliu) [inline src/utils.ts:1410]: Add comments for each added function.

File: src/utils.ts [hunk 3]
```diff
@@ -1405,4 +1405,97 @@ export function deserializeArgfromASM(contract: AbstractContract, arg: Argument,
...
+    }).join('');
+
+    return bsv.crypto.Hash.sha256(Buffer.from(jointbytes, 'hex')).toString('hex');
+  }
+}
+
+
+export function sortmap(map: Map<SingletonParamType, SingletonParamType>): Map<SingletonParamType, SingletonParamType> {
+  return new Map([...map.entries()].sort((a, b) => {
+    return BN.fromSM(Buffer.from(flattenSha256(a[0]), 'hex'), {
```
Comment (xhliu) [inline src/utils.ts:1462]: why `a[0]` only?

File: test/utils.test.ts [hunk 4]
```diff
@@ -1101,4 +1103,81 @@ describe('utils', () => {
...
   })
 
+
+  describe('flattenSha256() ', () => {
+
+    it('flattern data', () => {
+
+
+      expect(flattenSha256(1))
+        .to.be.equal("4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a");
```
Comment (xhliu) [inline test/utils.test.ts:1113]: How do we know these hashed values r correct? Combine them w/ .scrypt to compare what Typescript calculates w/ sCrypt.

**Review by xhliu — CHANGES_REQUESTED** [review]
💯 Looks good

File: src/utils.ts (hunk 3 above)
Comment (zhfnjust) [reply src/utils.ts:1462]: a is map entry, a[0]i is key, a[1] is value

File: src/compilerWrapper.ts [hunk 5]
```diff
@@ -571,6 +580,27 @@ export function getAliasDeclaration(astRoot, dependencyAsts): Array<AliasEntity>
   }).flat(1);
 }
 
+export function getGenericDeclaration(astRoot, dependencyAsts): Array<GenericEntity> {
```
Comment (xhliu) [inline src/compilerWrapper.ts:583]: Add comment for **ALL** non-trivial functions.

File: src/utils.ts [hunk 6]
```diff
@@ -544,6 +545,18 @@ export function findStructByName(name: string, s: StructEntity[]): StructEntity
...
 
+export function findLibraryByGeneric(type: string): string {
+
+  if(isGenericType(type)) {
+    const group = type.split('<');
+    const library = group[0].trim();
+    return library;
+  }
+
+  throw new Error(type + ' is not generic type');
```
Comment (xhliu) [inline src/utils.ts:556]: throw new Error(type + ' is not a generic type');

File: test/alias.test.ts [hunk 7]
```diff
@@ -164,13 +164,13 @@ describe('Alias type check', () => {
...
+      // expect(() => {
+      //   finalTypeResolver("Person")
+      // }).to.throw("typeResolver with unknown type Person")
 
-      expect(() => {
-        finalTypeResolver("Person[3]")
-      }).to.throw("typeResolver with unknown elemTypeName Person")
+      // expect(() => {
+      //   finalTypeResolver("Person[3]")
+      // }).to.throw("typeResolver with unknown elemTypeName Person")
```
Comment (xhliu) [inline test/alias.test.ts:173]: Delete dead code

File: src/utils.ts [hunk 8]
```diff
@@ -1487,7 +1501,7 @@ export function findKeyIndex(map: Map<SingletonParamType, SingletonParamType>, k
 }
 
 
-
+// serialize the map, but only flattenSha256 of the key and value
 export function toStorage(map: Map<SingletonParamType, SingletonParamType>): Bytes {
```
Comment (xhliu) [inline src/utils.ts:1505]: rename to `toData()`

File: test/fixture/mapTest.scrypt [hunk 9]
```diff
@@ -91,7 +91,7 @@ contract MapTest {
         require(map.set(7989009, 9061514, 1));
     }
 
-    public function testInsert(MapEntry[MAX] entries, bytes storage)// var length array
+    public function testInsert(MapEntry[MAX] entries, bytes data)// var length array
```
Comment (xhliu) [inline test/fixture/mapTest.scrypt:94]: remove all
`// var length array` everywhere

File: test/fixture/stateMap.scrypt [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+
+
+    public function insert(MapEntry entry, SigHashPreimage preimage)  {
+        require(Tx.checkPreimage(preimage));
+        HashedMap<int, int> map = new HashedMap(this._mpData);
+        int size = map.size();
+        require(!map.has(entry.key, entry.keyIndex));
+        map.set(entry.key, entry.val, entry.keyIndex);
+        require(map.size() == size + 1);
```
Comment (xhliu) [inline test/fixture/stateMap.scrypt:22]: Test `canGet()` as well.

`require(map.canGet())`

File: test/fixture/stateMap.scrypt [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        bytes output = Util.buildOutput(outputScript, Util.value(preimage));
+        require(hash256(output) == Util.hashOutputs(preimage));
+    }
+
+    public function update(MapEntry entry, SigHashPreimage preimage)  {
+        require(Tx.checkPreimage(preimage));
+        HashedMap<int, int> map = new HashedMap(this._mpData);
+        require(map.has(entry.key, entry.keyIndex));
+        map.set(entry.key, entry.val, entry.keyIndex);
+        this._mpData = map.data();
```
Comment (xhliu) [inline test/fixture/stateMap.scrypt:34]: add
`require(map.canGet())`

File: test/fixture/stateMap.scrypt [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        bytes outputScript = this.getStateScript();
+        bytes output = Util.buildOutput(outputScript, Util.value(preimage));
+        require(hash256(output) == Util.hashOutputs(preimage));
+    }
+
+    public function delete(int key, int keyIndex, SigHashPreimage preimage)  {
+        require(Tx.checkPreimage(preimage));
+        HashedMap<int, int> map = new HashedMap(this._mpData);
+        require(map.delete(key,  keyIndex));
```
Comment (xhliu) [inline test/fixture/stateMap.scrypt:43]:
```
require(map.has());
require(map.delete(key, keyIndex));
require(!map.has());
```

File: test/fixture/stateMap.scrypt [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        require(Tx.checkPreimage(preimage));
+        HashedMap<int, int> map = new HashedMap(this._mpData);
+        int size = map.size();
+        require(!map.has(entry.key, entry.keyIndex));
+        map.set(entry.key, entry.val, entry.keyIndex);
+        require(map.size() == size + 1);
+        this._mpData = map.data();
+        bytes outputScript = this.getStateScript();
+        bytes output = Util.buildOutput(outputScript, Util.value(preimage));
+        require(hash256(output) == Util.hashOutputs(preimage));
```
Comment (xhliu) [inline test/fixture/stateMap.scrypt:26]: Refactor: put common code into a function

-> require(passMap());

```
function passMap(...) : bool {
        this._mpData = map.data();
        bytes outputScript = this.getStateScript();
        bytes output = Util.buildOutput(outputScript, Util.value(preimage));
        return (hash256(output) == Util.hashOutputs(preimage));
}
```

File: test/fixture/stateMap.scrypt [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public function delete(int key, int keyIndex, SigHashPreimage preimage)  {
+        require(Tx.checkPreimage(preimage));
+        HashedMap<int, int> map = new HashedMap(this._mpData);
+        require(map.delete(key,  keyIndex));
+        this._mpData = map.data();
+        bytes outputScript = this.getStateScript();
+        bytes output = Util.buildOutput(outputScript, Util.value(preimage));
+        require(hash256(output) == Util.hashOutputs(preimage));
```
Comment (xhliu) [inline test/fixture/stateMap.scrypt:47]: -> require(passMap());

File: test/stateMapTest.test.ts [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+const outputAmount = inputSatoshis
+
+describe('test.stateMapTest', () => {
+    describe('stateMapTest', () => {
+        let mapTest, StateMapTest, MapEntry;
+
+        let map = new Map<number, number>();
+
+        before(() => {
+            const jsonDescr = loadDescription('stateMap_desc.json')
```
Comment (xhliu) [inline test/stateMapTest.test.ts:18]: dynamically compile in all tests since `*desc.json` are not committed.

File: test/stateMapTest.test.ts [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+            StateMapTest = buildContractClass(jsonDescr)
+            MapEntry = buildTypeClasses(StateMapTest).MapEntry;
+            mapTest = new StateMapTest(new Bytes(''))
```
Comment (xhliu) [inline test/stateMapTest.test.ts:21]: // empty initial map

File: test/stateMapTest.test.ts [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+                    satoshis: outputAmount
+                }))
+
+                const preimage = getPreimage(tx, mapTest.lockingScript, inputSatoshis)
+
+                mapTest.txContext = {
+                    tx: tx,
+                    inputIndex,
+                    inputSatoshis
+                }
```
Comment (xhliu) [inline test/stateMapTest.test.ts:93]: Same code as from Line 30 - 47. Put in a common function to avoid duplicate code.

**Review by xhliu — CHANGES_REQUESTED** [review]
👍 

A few improvements needed.

**Pull request merged**